common/djangoapps/util/model_utils.py

```diff
@@ -86,16 +85,63 @@ def clean_field(field_name, value):
     for field_name in changed_fields:
         if field_name not in excluded_fields:
-            tracker.emit(
-                event_name,
-                {
-                    "setting": field_name,
-                    'old': clean_field(field_name, changed_fields[field_name]),
-                    'new': clean_field(field_name, getattr(instance, field_name)),
-                    "user_id": user.id,
-                    "table": db_table
-                }
-            )
-    # Remove the now innacurate _changed_fields attribute.
-    if getattr(instance, '_changed_fields', None):
+            old_value = clean_field(field_name, changed_fields[field_name])
+            new_value = clean_field(field_name, getattr(instance, field_name))
+            emit_setting_changed_event(user, event_name, db_table, field_name, old_value, new_value)
+    # Remove the now inaccurate _changed_fields attribute.
+    if hasattr(instance, '_changed_fields'):
         del instance._changed_fields
+
+
+def emit_setting_changed_event(user, event_name, db_table, setting_name, old_value, new_value):
+    """Emits an event for a change in a setting.
+
+    Args:
+        user (User): the user that this setting is associated with.
+        event_name (str): the name of the event to be emitted.
+        db_table (str): the name of the table that we're modifying.
+        setting_name (str): the name of the setting being changed.
+        old_value (object): the value before the change.
+        new_value (object): the new value being saved.
+
+    Returns:
+        None
+    """
+    # Compute the maximum value length so that two copies can fit into the maximum event size
+    # in addition to all the other fields recorded.
+    max_value_length = settings.TRACK_MAX_EVENT / 4
+
+    serialized_old_value, old_was_truncated = _get_truncated_setting_value(old_value, max_length=max_value_length)
+    serialized_new_value, new_was_truncated = _get_truncated_setting_value(new_value, max_length=max_value_length)
+    truncated_values = []
+    if old_was_truncated:
+        truncated_values.append("old")
+    if new_was_truncated:
+        truncated_values.append("new")
+    tracker.emit(
+        event_name,
+        {
+            "setting": setting_name,
+            "old": serialized_old_value,
+            "new": serialized_new_value,
+            "truncated": truncated_values,
+            "user_id": user.id,
+            "table": db_table,
+        }
+    )
+
+
+def _get_truncated_setting_value(value, max_length=None):
+    """
+    Returns the truncated form of a setting value.
+
+    Returns:
+        truncated_value (object): the possibly truncated version of the value.
+        was_truncated (bool): returns true if the serialized value was truncated.
+    """
+    if isinstance(value, unicode) and max_length is not None and len(value) > max_length:
+        return u'{string}...'.format(string=value[0:max_length - 3]), True
```

Comment:
I wonder if we should really append "..." to the end. I think it would be less confusing with the ellipses, since the truncated variable makes it clear whether or not it was truncated. And the string may legitimately end with some number of periods.

@mulby, opinions?